keps/sig-autoscaling/5030-attach-limit-autoscaler/README.md

```diff
@@ -0,0 +1,722 @@
+
+
+### User Stories (Optional)
+
+#### Story 1
+- User has more than one pod that is pending because no existing node has any attach limit left.
+- Cluster autoscaler evaluates existing nodegroups.
+- It picks a nodegroup based on existing critireas and it accurately determines number of nodes it needs to spin up based on volumes that pending pods require.
+
+#### Story 2
+- A Kubernetes admin has one or more node where CSI driver is not installed. 
+- Without explicitly tainting the node or using node affinity in worklods, nodes which don't have CSI driver installed aren't used for scheduling pods that require volume.
+
+### Notes/Constraints/Caveats (Optional)
+
+Scheduler changes can only be merged after cluster-autoscaler changes has been GAed and there are no concerns about scheduler changes.
+
+### Risks and Mitigations
+
+As mentioned above, there is a risk associated with scheduler changes being merged before cluster-autoscaler changes can be merged, but we want to mitigate
+this by making sure cluster-autoscaler changes are merged first and has N+2 window before scheduler changes can be made GA. 
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+## Cluster Autoscaler changes
+
+We can split the implementation in cluster-autoscaler in two parts:
+- Scaling a node-group that already has one or more nodes.
+- Scaling a node-group that doesn’t have one or more nodes (Scaling from zero).
+
+### Scaling a node-group that already has one or more nodes.
+
+1. We propose a similar label as GPULabel added to the node that is supposed to come up with a CSI driver. This would ensure that, nodes which are supposed to have a certain CSI driver installed aren’t considered ready - https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/core/static_autoscaler.go#L979 until CSI driver is installed there.
+
+However, we also propose that a node will be considered ready as soon as corresponding CSI driver is being reported as installed via corresponding CSINode object.
+
+A node which is ready  but does not have CSI driver installed within certain time limit will be considered as NotReady and removed from the cluster.
+
+
+2. We propose that, we add volume limits and installed CSI driver information to framework.NodeInfo objects. So -
+
+```
+type NodeInfo struct {
+....
+....
+csiDrivers map[string]*DriverAllocatable
+..
+}
+
+type DriverAllocatable struct {
+    Allocatable int32
+}
+```
+
+3. We propose that, when saving `ClusterState` , we capture and add `csiDrivers` information to all existing nodes.
```

Comment:
Can we be more specific here? Does this mean listing CSINodes in `StaticAutoscaler.RunOnce()` and extending `ClusterSnapshot.SetClusterState()` to take a list of CSINodes?